File: 2023/Leetcode/problem695.py
class Solution:
    def __init__(self):
        self.WATER = 0


    def inBounds(self, grid, i, j):
        return 0 <= i < len(grid) and 0 <= j < len(grid[i])


    def dfs(self, grid, i, j, visited):
        if not self.inBounds(grid, i, j) or grid[i][j] == self.WATER or (i, j) in visited:
            return 0

        visited.add((i, j))

        up = self.dfs(grid, i + 1, j, visited)
        down = self.dfs(grid, i - 1, j, visited)
        right = self.dfs(grid, i, j + 1, visited)
        left = self.dfs(grid, i, j - 1, visited)

        return up + down + right + left + 1


    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        myMax = 0
        visited = set()

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] != self.WATER and (i, j) not in visited:
                    myMax = max(myMax, self.dfs(grid, i, j, visited))

        return myMax

File: 2023/Leetcode/problem695.java
class Solution {
    public final int WATER = 0;
    public final int LAND = 1;

    public int dfsArea(int[][] grid, int area, int i, int j) {
        grid[i][j] = WATER;

        if (j + 1 < grid[i].length && grid[i][j + 1] == LAND) {
            area = dfsArea(grid, area + 1, i, j + 1);
        }
        if (j - 1 >= 0 && grid[i][j - 1] == LAND) {
            area = dfsArea(grid, area + 1, i, j - 1);
        }
        if (i + 1 < grid.length && grid[i + 1][j] == LAND) {
            area = dfsArea(grid, area + 1, i + 1, j);
        }
        if (i - 1 >= 0 && grid[i - 1][j] == LAND) {
            area = dfsArea(grid, area + 1, i - 1, j);
        }

        return area;
    }

    public int maxAreaOfIsland(int[][] grid) {
        int currArea;
        int maxArea = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == LAND) {
                    currArea = dfsArea(grid, 1, i, j);
                    if (currArea > maxArea) {
                        maxArea = currArea;
                    }
                }
            }
        }

        return maxArea;
    }
}